apps/condo/domains/billing/schema/SumBillingReceiptsService.js

```diff
@@ -4,72 +4,79 @@
     },
 }
 
 const SumBillingReceiptsService = new GQLCustomSchema('SumBillingReceiptsService', {
-
     types: [
+        {
+            access: true,
+            type: 'input BillingReceiptsSumInput { period: String!, organizationId: ID, tin: String, importRemoteSystem: String }',
+        },
         {
             access: true,
             type: 'type BillingReceiptsSumOutput { sum: String! }',
         },
     ],
-    
     queries: [
         {
             access: access.canSumBillingReceipts,
-            schema: '_allBillingReceiptsSum (where: BillingReceiptWhereInput!): BillingReceiptsSumOutput',
-            resolver: async (parent, args, context, info, extra = {}) => {
-                const { where } = args
+            schema: '_allBillingReceiptsSum (where: BillingReceiptsSumInput!): BillingReceiptsSumOutput',
+            resolver: async (parent, args) => {
+                const { where: { period, tin, organizationId, importRemoteSystem } } = args
 
-                // We do not want to give user an easy way to query whole database. You must specify period in where query.
-                //
-                // It was done like this, and not as a separate argument for several reasons:
-                // 1. in order to keep validation of period in one place - in GraphQL
-                // 2. in order to keep all<model> styled api arguments somewhat consistent (e.x _allPaymentsSum / allBillingReceipts )
-                //
-                // If you try to add more periods to the where clause, like: where: (period=a1 period=a2) you will get an error
-                // If you try to be hacky and do this: where: (period=a1 period_gt=a1) you are going to have intersection of these clauses which means you will get only period=a1 ones!
-                if (!where.period || typeof where.period !== 'string') {
-                    throw new GQLError(ERRORS.PERIOD_BADLY_SPECIFIED)
+                if (!tin && !organizationId) {
+                    throw new GQLError(ERRORS.TIN_OR_ORGANIZATION_ID_MUST_BE_SPECIFIED)
                 }
 
-                const billingReceiptsLoader = new GqlWithKnexLoadList({
-                    listKey: 'BillingReceipt',
-                    fields: 'id toPay',
-                    where,
-                    sortBy: 'createdAt_DESC',
-                })
-                const objs = await billingReceiptsLoader.load()
-                const toPaySum = objs.reduce((accumulator, receipt) => {
-                    return accumulator.plus(new Big(receipt.toPay))
-                }, new Big(0))
+                const { keystone } = getSchemaCtx('BillingReceipt')
+                const { knex } = getDatabaseAdapter(keystone)
+
+                const query = knex('BillingReceipt')
+                    .sum('BillingReceipt.toPay as totalToPay')
+                    .innerJoin('BillingIntegrationOrganizationContext', 'BillingReceipt.context', 'BillingIntegrationOrganizationContext.id')
```

Comment:
JOINS might be expensive too.. Maybe we can filter out required contexts using our business logic and then using these contexts just sum billing receipts?

Reply:
Might be worth a try, might be not

Reply:
Agree, remove additional joins from BillingReceipts query.
Change it to separate query for contexts and filter BillingReceipts by context ids